### ISSUE TITLE
Error 403 when getting playplay license

### ISSUE THREAD

**Issue by glomatico** [glomatico/votify#5]
```
python -m votify "https://open.spotify.com/track/18gqCQzqYb0zvurQPlRkpo" -n
[INFO     08:31:31] Starting Votify
[INFO     08:31:33] (URL 1/1) Checking "https://open.spotify.com/track/18gqCQzqYb0zvurQPlRkpo"
[INFO     08:31:33] (Track 1/1 from URL 1/1) Downloading "Never Gonna Give You Up - 2022 Remaster"
[ERROR    08:31:35] (Track 1/1 from URL 1/1) Failed to download "Never Gonna Give You Up - 2022 Remaster"
Traceback (most recent call last):
  File "C:\Users\Glomatico\Documents\GitHub\votify\votify\utils.py", line 8, in check_response
    response.raise_for_status()
  File "C:\Users\Glomatico\Documents\GitHub\votify\.venv\Lib\site-packages\requests\models.py", line 1024, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 403 Client Error: Forbidden for url: https://gew4-spclient.spotify.com/playplay/v1/key/b7098e5bb8c2c1639bb44e0c17549e6d9f24bb3f

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\Glomatico\Documents\GitHub\votify\votify\cli.py", line 584, in main
    downloader_song.download(
  File "C:\Users\Glomatico\Documents\GitHub\votify\votify\downloader_song.py", line 137, in download
    self._download(*args, **kwargs)
  File "C:\Users\Glomatico\Documents\GitHub\votify\votify\downloader_song.py", line 213, in _download
    decryption_key = self.get_decryption_key(stream_info)
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Glomatico\Documents\GitHub\votify\votify\downloader_audio.py", line 94, in get_decryption_key
    decryption_key = self.downloader.get_playplay_decryption_key(
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Glomatico\Documents\GitHub\votify\votify\downloader.py", line 376, in get_playplay_decryption_key
    playplay_license_response_bytes = self.spotify_api.get_playplay_license(
                                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Glomatico\Documents\GitHub\votify\votify\spotify_api.py", line 283, in get_playplay_license
    check_response(response)
  File "C:\Users\Glomatico\Documents\GitHub\votify\votify\utils.py", line 10, in check_response
    _raise_response_exception(response)
  File "C:\Users\Glomatico\Documents\GitHub\votify\votify\utils.py", line 14, in _raise_response_exception
    raise Exception(
Exception: Request failed with status code 403:
[INFO     08:31:35] Done (1 error(s))
```

**Comment by xMidNightFTW** [comment]
```python
C:\Users\xMidNight>votify "https://open.spotify.com/track/71MEWIf4Itr8Pd9OW5Mq7q?si=e90e7b244a0442d2"
[INFO     10:23:17] Starting Votify
[INFO     10:23:18] (URL 1/1) Checking "https://open.spotify.com/track/71MEWIf4Itr8Pd9OW5Mq7q?si=e90e7b244a0442d2"
[INFO     10:23:18] (Track 1/1 from URL 1/1) Downloading "Hola Señorita"
[ERROR    10:23:19] (Track 1/1 from URL 1/1) Failed to download "Hola Señorita"
Traceback (most recent call last):
  File "C:\Users\xMidNight\AppData\Local\Programs\Python\Python311\Lib\site-packages\votify\utils.py", line 8, in check_response
    response.raise_for_status()
  File "C:\Users\xMidNight\AppData\Local\Programs\Python\Python311\Lib\site-packages\requests\models.py", line 1021, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 403 Client Error: Forbidden for url: https://gew4-spclient.spotify.com/playplay/v1/key/c9931ec76d7388d78a1e1f65af41e3197f5a7cd9

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\xMidNight\AppData\Local\Programs\Python\Python311\Lib\site-packages\votify\cli.py", line 584, in main
    downloader_song.download(
  File "C:\Users\xMidNight\AppData\Local\Programs\Python\Python311\Lib\site-packages\votify\downloader_song.py", line 137, in download
    self._download(*args, **kwargs)
  File "C:\Users\xMidNight\AppData\Local\Programs\Python\Python311\Lib\site-packages\votify\downloader_song.py", line 213, in _download
    decryption_key = self.get_decryption_key(stream_info)
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\xMidNight\AppData\Local\Programs\Python\Python311\Lib\site-packages\votify\downloader_audio.py", line 94, in get_decryption_key
    decryption_key = self.downloader.get_playplay_decryption_key(
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\xMidNight\AppData\Local\Programs\Python\Python311\Lib\site-packages\votify\downloader.py", line 370, in get_playplay_decryption_key
    playplay_license_response_bytes = self.spotify_api.get_playplay_license(
                                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\xMidNight\AppData\Local\Programs\Python\Python311\Lib\site-packages\votify\spotify_api.py", line 283, in get_playplay_license
    check_response(response)
  File "C:\Users\xMidNight\AppData\Local\Programs\Python\Python311\Lib\site-packages\votify\utils.py", line 10, in check_response
    _raise_response_exception(response)
  File "C:\Users\xMidNight\AppData\Local\Programs\Python\Python311\Lib\site-packages\votify\utils.py", line 14, in _raise_response_exception
    raise Exception(
Exception: Request failed with status code 403:
[INFO     10:23:19] Done (1 error(s))
```

**Comment by glomatico** [comment]
Should be fixed on 1.4.2

**Comment by glomatico** [comment]
welp

**Comment by mbucc9000** [comment]
Yep. Also happening to me :(

**Comment by glomatico** [comment]
Fixed by updating to the latest version of re-playplay
```bash
pip install re-unplayplay -U
```

**Comment by mbucc9000** [comment]
That worked! I was also using [Zotify](https://github.com/zotify-dev/zotify) and was ALSO getting errors... it seems like updating re-playplay fixed my issue there as well. 

Appreciate the work!

**Comment by 1967antonio** [comment]
some problem

**Comment by xyloflake** [comment]
Can confirm it's back again 
![image](https://github.com/user-attachments/assets/9c17075f-7609-4dd4-89b4-72ec861e5d08)

The token needs to be updated.

**Comment by es3n1n** [comment]
Fixed in re-unplayplay@v1.2.6. `python3 -m pip install -U re-unplayplay`

**Comment by Abhi5033** [comment]
> Fixed in re-unplayplay@v1.2.6. `python3 -m pip install -U re-unplayplay`

I updated the reunplayplay but still receiving 403 error

![image](https://github.com/user-attachments/assets/624c4c6a-9ab1-4d27-a3a4-b5d91c67e2f8)

![image](https://github.com/user-attachments/assets/d6ae3a1e-8441-415f-8ebd-1cfb84bb2ec5)

Please help

**Comment by jaguarondi** [comment]
I have exactly the same error:
re-unplayplay==1.2.6
votify==1.4.4

**Comment by xyloflake** [comment]
@es3n1n token got banned again?

**Comment by es3n1n** [comment]
> @es3n1n token got banned again?

I think so, yeah

**Comment by xyloflake** [comment]
> I think so, yeah

That was unreasonably quick....

**Comment by es3n1n** [comment]
Fixed in re-unplayplay@v1.2.7. `python3 -m pip install -U re-unplayplay`
You will also need to wait until #20 is merged

**Comment by es3n1n** [comment]
Nice, new tokens got banned in less than an hour. 
s/o to the spotify employees that are following our updates ❤️

**Comment by es3n1n** [comment]
Fixed in re-unplayplay@v1.2.8. `python3 -m pip install -U re-unplayplay`

**Comment by es3n1n** [comment]
Banned in 27 minutes, yeah no thanks

**Comment by xyloflake** [comment]
> Banned in 27 minutes, yeah no thanks

What the hell is happening

**Comment by xyloflake** [comment]
> Banned in 27 minutes, yeah no thanks

These tokens that you are using, how old are these builds from?

**Comment by xyloflake** [comment]
If we use a new enough token, they can't ban them, now can they? They'd be fundamentally breaking over a million devices if they ban the latest token(?).

**Comment by paul19920801** [comment]
any idea how to make your own token...

I would rather make my own than deal with thinks stop working...

**Comment by LeonSolisPedro** [comment]
Ohh! time to find other alternatives

@xyloflake

**Comment by s0llvan** [comment]
> Ohh! time to find other alternatives
> 
> @xyloflake

if you have any suggestions...

**Comment by xyloflake** [comment]
> any idea how to make your own token...
> 
> I would rather make my own than deal with thinks stop working...

Reverse engineer the binaries as instructed above @s0llvan

**Comment by s0llvan** [comment]
> > any idea how to make your own token...
> > I would rather make my own than deal with thinks stop working...
> 
> Reverse engineer the binaries as instructed above @s0llvan

Existing suggestions lol, not starting from scratch

**Comment by ypraag-cloudinary** [comment]
Bummer, same issue `re-unplayplay=1.2.8`
```
[INFO     12:06:28] Starting Votify
[INFO     12:06:29] (URL 1/1) Checking "https://open.spotify.com/track/19KlZwqlT3fguP2BeHF1Q1?si=d5a641b5f8764d9c"
[INFO     12:06:29] (Track 1/1 from URL 1/1) Downloading "Disease"
[ERROR    12:06:30] (Track 1/1 from URL 1/1) Failed to download "Disease"
Traceback (most recent call last):
  File "/opt/homebrew/lib/python3.11/site-packages/votify/utils.py", line 8, in check_response
    response.raise_for_status()
  File "/opt/homebrew/lib/python3.11/site-packages/requests/models.py", line 1024, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 403 Client Error: Forbidden for url: https://gew4-spclient.spotify.com/playplay/v1/key/0a1d62ff7d1715318f1704e7f2baad1b775b2b3e

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/homebrew/lib/python3.11/site-packages/votify/cli.py", line 597, in main
    downloader_song.download(
  File "/opt/homebrew/lib/python3.11/site-packages/votify/downloader_song.py", line 137, in download
    self._download(*args, **kwargs)
  File "/opt/homebrew/lib/python3.11/site-packages/votify/downloader_song.py", line 213, in _download
    decryption_key = self.get_decryption_key(stream_info)
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/votify/downloader_audio.py", line 98, in get_decryption_key
    decryption_key = self.downloader.get_playplay_decryption_key(
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/votify/downloader.py", line 381, in get_playplay_decryption_key
    playplay_license_response_bytes = self.spotify_api.get_playplay_license(
                                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/votify/spotify_api.py", line 287, in get_playplay_license
    check_response(response)
  File "/opt/homebrew/lib/python3.11/site-packages/votify/utils.py", line 10, in check_response
    _raise_response_exception(response)
  File "/opt/homebrew/lib/python3.11/site-packages/votify/utils.py", line 14, in _raise_response_exception
    raise Exception(
Exception: Request failed with status code 403: 
[INFO     12:06:30] Done (1 error(s))

```

**Comment by xyloflake** [comment]
> Existing suggestions lol, not starting from scratch

No existing solutions working atm @s0llvan 

Spotify is - also for some reason breaking all the mods. I think the war has begun.

**Comment by darkblackx** [comment]
As a software engineer I understand why things are broken at the moment. I appreciate the effort made to make this project work at all. Recent changes by Spotify stopped  this 'solution' from working. Maybe its temporary, maybe its permanent. 
I think the main 'readme' of this project should reflect that it does not work, and if the project/team is working on a solution.

No harm in updating the readme to state that an upstream change broke the current solution. Prevents multiple issues being reported, which devs and users need to read though to understand. 

### Rant about issues reported 
In the last few years too many 'regular folks' have started raising github issues for their own support. Can't blame them, and the fact they got that far is an achievement! However, it increases the 'noise' for project devs, and more issues for everyone to read through to see if the issue they are facing is already logged....A non tech user that make it this far will just open a new issue, even though its a duplicate. The result is some solutions / suggestions communicated in one issue, but not the other...[ For the pedantic: In an ideal world duplicate issues would be merged/ handled etc. however this would be a side project for all devs involved, after their paying day jobs. its highly unlikely they want to use their free time organising and dealling with 'issues' reported that actual issues. If your able to identify duplicate issues, and or issues that are likely caused by the individual reporter, and not at project issue, I'm sure the project would love any time you can spare to sort 'issues' raised. Reach out.

I know the pressure is on to 'fix' the solution, and also provide a deadline for when it will be fixed from ungrateful users...
Again, I suggest updating the readme so that its clear that things are currently 'broken' due to a upstream change to prevent duplicate 'issues' raised, and users of all types having to read through all open issues to see if there is a current problem. In the same notice, remind people this project only exists as a passion project, and all work is done by unpaid volunteers. Link to ways to donate/support the effort.

**TL; DR;** Please have a project 'state' which shows when spotify/upstream has changed something, preventing this solution from working. Create one issue to track the situation. Link any user issues to the master 'issue' and provide updates in the one place. This would prevent many people wasting individual time trying to get this solution working, when it won't, and its known it won't. 'You' don't have to provide an ETA for a fix, but 'users' will use all available avenues to ask for one. 

I recommend vague statements which give some sort of timeline, even if that that timeline is infinite. ie; 'a upstream change as blocked part of our process, but we can see a work around, should be available shortly 12-24 hours', or, 'an upstream change has stopped this solution from working due to an encoding change. This will take significant time and effort to work around, no eta can be provided. The project team would welcome engineers with experience in xyz to aid.

Yes, this 'help' (probably unwanted) was too long, but I hope at least one person read it. Imagine if all software/code had explainations like the above... 

(probably best if us software engineers continue the path of no comments or documentation. It results it more billible hours and jobs!)

**Comment by PwLDev** [comment]
I've found a working payload for PlayPlay license requests! (It's in JS btw)

Endpoint: https://spclient.wg.spotify.com/playplay/v1/key/{fileId}
Headers:
"Authorization": `Bearer ${accessToken}`,
"Accept": "*"
Base64 encoded payload: `CAMSEAGny+DVFTUfacKr9zszemsgASgBMOfG9I8G`
Payload source: https://github.com/DaXcess/node-spotify-ap/blob/81e66fd5122936f299118fd75fd47420076f12e8/lib/audio/AudioKeyManager.ts#L84

Tried this in my project and as of today, since 3 days ago, it's working with no issues, always receiving 200 with correct Protobuf response

Now the problem with this is that I haven't found yet a deobfuscation algorithm that utilizes this token from the payload. Re-playplay is deobfuscating the key incorrectly resulting in partial decryption.
Btw: sorry if my English is bad, still practicing tho.
